Support cookies

Kas-tle opened the pull request:
Closes #37

Kas-tle committed:
Support cookies

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle reviewed (commented):
Tested with Geyser and seems to connect fine. Verified cookie with wireshark. (Note if you try this yourself I also had to do a PR to wireshark to fix its handling of encryption being changed on the second reply, so see that PR here https://gitlab.com/wireshark/wireshark/-/merge_requests/15123 and build can be downloaded here https://github.com/Kas-tle/wireshark/actions/runs/8547183545)

Kas-tle commented:
How far back would you need to know? I'll do my best to check as far as I can but unsure how far back the unofficial launcher can go.

Alemiz112 commented:
I believe have support back to raknet version 8 or maybe 7. Let me know which versions you were able to test with. I should be able to test with the very old versions later.

Alemiz112 commented on transport-raknet/src/main/java/org/cloudburstmc/netty/channel/raknet/config/RakChannelOption.java:
```diff
@@ -155,6 +155,12 @@ public class RakChannelOption<T> extends ChannelOption<T> {
+    /**
+     * Whether to send a cookie to the client during the connection process.
+     */
+    public static final ChannelOption<Boolean> RAK_SEND_COOKIE =
```
Dont forget to add this into `DefaultRakServerConfig#getOptions()`.

Kas-tle replied:
https://github.com/CloudburstMC/Network/pull/38/files#diff-741fd5ee99ef49922ba03b8640c20493d5b2348e979747d1dd7748fdcca44be8R61

Alemiz112 commented on transport-raknet/src/main/java/org/cloudburstmc/netty/handler/codec/raknet/client/RakClientOfflineHandler.java:
```diff
@@ -153,24 +154,32 @@ protected void channelRead0(ChannelHandlerContext ctx, ByteBuf buf) throws Excep
 
         this.state = RakOfflineState.HANDSHAKE_2;
-        this.sendOpenConnectionRequest2(ctx.channel());
+        if (this.security) {
+            this.sendOpenConnectionRequest2(ctx.channel(), this.cookie);
+        } else {
+            this.sendOpenConnectionRequest2(ctx.channel());
+        }
```
Do we need two methods for this?

Kas-tle replied:
Hmm yeah probably not.

Kas-tle replied:
https://github.com/CloudburstMC/Network/pull/38/files#diff-10e14df5630dd76f7375a6a7764a7d8f6d739a8ed5097563b0492597502ab014L199-R214

Alemiz112 commented on transport-raknet/src/main/java/org/cloudburstmc/netty/handler/codec/raknet/server/RakServerOfflineHandler.java:
```diff
@@ -58,6 +61,12 @@ public class RakServerOfflineHandler extends AdvancedChannelInboundHandler<Datag
             .build();
 
+    private final ExpiringMap<InetSocketAddress, Integer> cookies = ExpiringMap.builder()
+            .expiration(10, TimeUnit.SECONDS)
```
Can we merge this with `pendingConnections` map?

Kas-tle replied:
Yes already planning to do this based on feedback from Supreme. Just unsure of how to handle the case of no pending connection equating to invalid version. Imo the correct thing to do for both cases is to just not reply but not sure if I should keep the old logic when send cookie is disabled.

Alemiz112 replied:
I initially used `sendIncompatibleVersion` because there isn't any other packet to properly close connection and show some message on client side. 
I would always use `pendingConnections.remove` in `onOpenConnectionRequest2` to get the entry. If it returns `null`, I would send wrong version because no version was set in OpenConReq1 and if there is cookie mismatch, just return, but make sure entry is also removed from  `pendingConnections`, so whole sequence is enforced to restart.

Kas-tle replied:
But why would the vanilla client ever send a packet with no protocol version, and if it isn't a vanilla client, then why should we bother?

Kas-tle replied:
Though I guess it could be that they somehow missed the first two but that seems kinda sus.

Alemiz112 replied:
I have seen vanilla resending OpenConReq2 multiple times with older versions. But I guess we can also just ignore it and if client resends it for whatever reason, it should be able to deal with it.

Kas-tle replied:
https://github.com/CloudburstMC/Network/pull/38/files#diff-5679517bda5af958106b1a7007f151e9323b1c2e56c1b8ab7ceeb8e7a71d1e4a

Kas-tle committed:
Address existing GH reviews

- Add RAK_SEND_COOKIE to DefaultRakServerConfig#getOptions()
- Consolidate RakClientOfflineHandler#sendOpenConnectionRequest2 to single method
- Combine pendingConnections ExpiringMap

Signed-off-by: GitHub <noreply@github.com>

Kas-tle committed:
Merge remote-tracking branch 'upstream/develop' into feature/cookie

Signed-off-by: GitHub <noreply@github.com>

Kas-tle commented:
Will implement optimal SecureRandom later tonight.

Kas-tle committed:
Remove unused packages

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Create provider for preferred SecureRandom algorithm

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Camotoy commented on transport-raknet/src/main/java/org/cloudburstmc/netty/util/SecureAlgorithmProvider.java:
```diff
@@ -0,0 +1,38 @@
+public class SecureAlgorithmProvider {
+    private static final String SECURITY_ALGORITHM;
+    // SecureRandom algorithms in order of most preferred to least preferred.
+    private static List<String> preferredAlgorithms = Arrays.asList(
```
This doesn't have to be static if you just make it null later? Why not create it within the static block?

Kas-tle replied:
https://github.com/CloudburstMC/Network/pull/38/files#diff-6094c732dc558796d1151bc5a36550b609312bb130fd0fc8cd84d77c9984f9d1

Kas-tle committed:
preferredAlgorithms in block

Kas-tle committed:
Why we should not do web edits

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Alemiz112 reviewed (approved):
Thanks for this feature!
